static/js/components/dashboard/CourseRow.js

```diff
@@ -1,56 +1,163 @@
+    run.status !== STATUS_MISSED_DEADLINE
+  );
+
+  needsToEnrollAgain = (run: CourseRun): boolean => (
+    run.status === STATUS_MISSED_DEADLINE || run.status === STATUS_NOT_PASSED
+  );
+
+  futureEnrollableRun = (course: Course): CourseRun|null => (
+    (course.runs.length > 1 && course.runs[1].status === STATUS_OFFERED) ? course.runs[1] : null
+  );
+
+  pastCourseRuns = (course: Course): Array<CourseRun> => (
+    (course.runs.length > 1) ?
+      R.drop(1, course.runs).filter(run => run.status !== STATUS_OFFERED) :
+      []
+  );
+
+  getFirstRun(): CourseRun|null {
+    const { course } = this.props;
+    let firstRun: CourseRun = {};
+    if (course.runs.length > 0) {
+      firstRun = course.runs[0];
+    }
+    return firstRun;
+  }
+
+  renderRowColumns(run: CourseRun): Array<React$Element<*>> {
     const {
       course,
       coursePrice,
+      checkout,
+      now,
       financialAid,
       hasFinancialAid,
-      checkout,
       openFinancialAidCalculator,
-      now,
       addCourseEnrollment,
     } = this.props;
 
-    return <Grid className="course-row">
-      <Cell col={6}>
-        <CourseDescription course={course} />
-      </Cell>
-      <Cell col={2}>
-        <CourseGrade course={course}/>
+    let lastColumnSize, lastColumnClass;
+    let columns = [
+      <Cell col={6} key="1">
+        <CourseDescription courseRun={run} courseTitle={course.title} />
       </Cell>
-      <Cell col={4}>
+    ];
+
+    if (this.shouldDisplayGradeColumn(run)) {
+      columns.push(
+        <Cell col={2} key="2">
+          <CourseGrade courseRun={run} />
+        </Cell>
+      );
+      lastColumnSize = 4;
+    } else {
+      lastColumnSize = 6;
+      lastColumnClass = 'long-description';
+    }
+
+    columns.push(
+      <Cell col={lastColumnSize} className={lastColumnClass} key="3">
         <CourseAction
-          course={course}
+          courseRun={run}
           coursePrice={coursePrice}
+          checkout={checkout}
+          now={now}
           hasFinancialAid={hasFinancialAid}
           financialAid={financialAid}
-          checkout={checkout}
           openFinancialAidCalculator={openFinancialAidCalculator}
-          now={now}
           addCourseEnrollment={addCourseEnrollment}
         />
       </Cell>
-    </Grid>;
+    );
+    return columns;
+  }
+
+  renderSubRows(): Array<React$Element<*>> {
+    const {
+      course,
+      coursePrice,
+      checkout,
+      now,
+      financialAid,
+      hasFinancialAid,
+      openFinancialAidCalculator,
+      addCourseEnrollment,
+    } = this.props;
+
+    let firstRun = this.getFirstRun();
+
+    let subRows = [];
+    let subRowRuns = [];
+
+    if (this.needsToEnrollAgain(firstRun)) {
+      subRowRuns.push(this.futureEnrollableRun(course));
+    }
+    let completedRuns = this.pastCourseRuns(course);
+    if (completedRuns.length > 0) {
+      subRowRuns = subRowRuns.concat(completedRuns);
+    }
+
+    if (subRowRuns.length > 0) {
```

Comment:
This check is also unnecessary since the for loop below won't do anything on an empty list